src/lib/network_pool/transaction_pool.ml

```diff
@@ -1273,12 +1273,25 @@ struct
             ~f:(fun _ ->
               Counter.inc_one Transaction_pool.transactions_added_to_pool )) ;
         (* partition the results *)
-        let accepted, rejected =
-          List.partition_map add_results ~f:(function
+        let all_drops =
+          List.fold add_results ~init:Sequence.empty ~f:(fun acc result ->
+              match result with
+              | Ok (_cmd, dropped) ->
+                  Sequence.append acc dropped
+              | Error _ ->
+                  acc )
+        in
+        let accepted, rejected, _dropped =
+          List.partition3_map add_results ~f:(function
             | Ok (cmd, _dropped) ->
-                Either.First cmd
+                if
+                  Sequence.mem all_drops cmd
```

Comment:
`Sequence.mem` is `O(n)`. Since this is in a loop over the results from adding, we should use an `O(1)` solution here. I'd recommend collecting the transaction hashes of dropped commands into a set and checking membership against that. The transactions hashes should already be available at this point in the code.